specification/dartLangSpec.tex

```diff
@@ -2978,16 +2978,26 @@ \subsubsection{Generative Constructors}
 \end{grammar}
 
 \LMHash{}%
-A \Index{constructor parameter list} is a parenthesized, comma-separated list of formal constructor parameters.
-A \Index{formal constructor parameter} is either a formal parameter (\ref{formalParameters}) or an initializing formal.
-An \Index{initializing formal} has the form \code{\THIS{}.\id}, where \id{} is the name of an instance variable of the immediately enclosing class.
-It is a compile-time error if \id{} is not an instance variable of the immediately enclosing class.
-It is a compile-time error if an initializing formal is used by a function other than a non-redirecting generative constructor.
-
-\LMHash{}%
-If an explicit type is attached to the initializing formal, that is its static type.
-Otherwise, the type of an initializing formal named \id{} is $T_{id}$, where $T_{id}$ is the type of the instance variable named \id{} in the immediately enclosing class.
-It is a compile-time error if the static type of \id{} is not a subtype of $T_{id}$.
+A \Index{constructor parameter list} is
+a parenthesized, comma-separated list of formal constructor parameters.
+A \Index{formal constructor parameter} is a formal parameter
+(\ref{formalParameters}),
+which can be an initializing formal.
+An \Index{initializing formal} contains a term of the form \code{\THIS.\id},
+where \id{} is the name of the initializing formal.
+It is a compile-time error if \id{} is not also the name of
+an instance variable of the immediately enclosing class.
+It is a compile-time error if an initializing formal is declared in
+a function declaration which is not a non-redirecting generative constructor.
+
+\LMHash{}%
+Assume that $p$ is a declaration of an initializing formal parameter named \id.
+Let $T_{id}$ be the type of the instance variable named \id{} in
+the immediately enclosing class.
+If $p$ has a type annotation $T$ then the declared type of $p$ is $T$.
+Otherwise, the declared type of $p$ is $T_{id}$.
+It is a compile-time error if the declared type of $p$
+is not a subtype of $T_{id}$.
```

Comment:
(I it worth mentioning that a constructor parameter cannot be `covariant`, so "subtype of" is sufficient).

Reply:
Interesting corner! ;)

Actually, I'd say that we need to ensure that it would not be an error to use an expression whose type is the type of the formal in an initialization of the instance variable; next, we don't want `dynamic`, so the formal type must be a subtype of the variable type. If the instance variable is declared to be covariant then an overriding setter can have a more special parameter type, but that shouldn't matter, because that overriding setter would not be invoked in order to initialize the variable in this class anyway. So we can rely on the declared type of the variable in this class even when the variable is covariant.